AU2/database/model/Assassin.py

```diff
@@ -42,9 +58,104 @@ def __post_init__(self):
+        which may differ from when the pseudonym was physically entered into AutoUmpire.
+        The start of validity may be `None`, which represents that the pseudonym is always valid.
+
+        :param i: the index of the pseudonym to get the validity of
+        :return: `None` if the relevant pseudonym is always valid, otherwise a `datetime.datetime` object representing when this assassin was granted the pseudonym
+        """
+        return self.pseudonym_datetimes[i] if i in self.pseudonym_datetimes else None
+
+    def add_pseudonym(self, val: str, valid_from: dt.datetime = get_now_dt()) -> int:
+        """
+        Adds a new pseudonym to this assassin.
+
+        :param val: text of new pseudonym
+        :param valid_from: start of validity of new pseudonym; defaults to the current datetime, may be explicitly set to `None` to indicate that the pseudonym is always valid
+        :return: index of the newly-added pseudonym
+        """
+        new_i = len(self.pseudonyms)
+        if isinstance(valid_from, dt.datetime):
+            self.pseudonym_datetimes[new_i] = valid_from
+        elif valid_from is not None:
+            raise TypeError("add_pseudonym expects a datetime.datetime object or None in the second argument")
+        self.pseudonyms.append(val)
+        return new_i
+
+    def edit_pseudonym(self, i: int, new_val: str, new_valid_from: dt.datetime):
+        """
+        Edits a pseudonym, referenced by index, from this assassin.
+        Both the new text and new start of validity must be specified.
+        (This is to avoid confusion between using `None` as a placeholder default value,
+        and `None` representing an always-valid pseudonym).
+
+        :param i: the index of the pseudonym to edit
+        :param new_val: the new text of the pseudonym
+        :param new_valid_from: the new start of validity of the pseudonym; may be explicitly set to `None` to indicate the pseudonym should always be valid
+        """
+        if new_val.strip() == "":
+            raise ValueError("Cannot set a blank pseudonym.")
+
+        if new_valid_from is not None:
+            if i == 0:
+                raise ValueError("Cannot set a validity for the initial pseudonym.")
+            else:
+                self.pseudonym_datetimes[i] = new_valid_from
+        # a validity of `None`, which represents the pseudonym being valid forever
+        elif i in self.pseudonym_datetimes:
+            del self.pseudonym_datetimes[i]
+
+        self.pseudonyms[i] = new_val
+
+    def delete_pseudonym(self, i: int):
+        """
+        Deletes a pseudonym, referenced by index, from this assassin.
+        Internally, the pseudonym is set to "" so that the indices of other pseudonyms are unchanged.
+        The initial pseudonym (i=0) cannot be deleted.
+
+        :param i: the index of the pseudonym to delete
+        """
+        if i == 0:
+            raise ValueError("Cannot delete initial pseudonym.")

```

Comment:
Hmm... Yes I think it is probably better for the program not to crash over a silly error. My one concern is that it may make debugging more time-consuming as the source of weird behaviour will be less clear than if the program just raises an exception. I'm comfortable with it possibly failing silently in this case though since the "weird" behaviour here is just a guardrail against the umpire doing something silly.

Just to clarify: this `Try` doesn't already exist and would need to be declared somewhere in AU2? I can't find any reference to it anywhere.